### PR TITLE
 prio nodes farmerbot

### PR DESCRIPTION
**Pull request by rawdaGastan**
### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-go/issues/978

### CODE REVIEW

**Comment by rawdaGastan** [comment]
Should we use a slice of nodes instead of a map?

File: farmerbot/README.md [hunk 1]
```diff
@@ -99,6 +99,8 @@ included_nodes: [optional, if no nodes are added then the farmerbot will include
   - "<your node ID to be included, required at least 2>"
 excluded_nodes:
   - "<your node ID to be excluded, optional>"
+priority:
```
Comment (Eslam-Nawara) [inline farmerbot/README.md]: It should be `priority_nodes`

File: farmerbot/internal/power.go [hunk 2]
```diff
@@ -148,7 +148,15 @@ func calculateResourceUsage(nodes map[uint32]node) (uint64, uint64) {
...
-	for nodeID, node := range f.nodes {
+	nodesKeys := make([]uint32, 0, len(f.nodes))
+	for k := range f.nodes {
+		nodesKeys = append(nodesKeys, k)
+	}
+
+	for i := 0; i < len(f.nodes); i++ {
+		nodeID := nodesKeys[i]
+		node := f.nodes[nodeID]
+
```
Comment (Eslam-Nawara) [inline farmerbot/internal/power.go]: Is there a reason this is done this way? I mean, I think this does the same job as before

**Comment by Eslam-Nawara** [comment]
> Should we use a slice of nodes instead of a map?

I think yes, looping over map, saving the keys in array then depending on the order of the keys this won't guarantee the order of the keys to be as expected either

File: farmerbot/internal/state.go [hunk 3]
```diff
@@ -212,55 +212,68 @@ func getNode(
...
 }
 
 // UpdateNode updates a node in the config
 func (s *state) updateNode(node node) error {
-	s.m.Lock()
-	defer s.m.Unlock()
-
-	nodeID := uint32(node.ID)
+	// s.m.Lock()
+	// defer s.m.Unlock()
```
Comment (Eslam-Nawara) [inline farmerbot/internal/state.go]: I think the lock should be added as we're updating the list

File: farmerbot/internal/farmerbot.go [hunk 4]
```diff
@@ -328,11 +333,39 @@ func (f *FarmerBot) iterateOnNodes(ctx context.Context, subConn Substrate) error
 	return nil
 }
 
+func addPriorityToNodes(priorityNodes, farmNodes []uint32) []uint32 {
+	updatedFarmNodes := make([]uint32, len(farmNodes))
+	copy(updatedFarmNodes, farmNodes)
+
+	updatedPriorityNodes := make([]uint32, len(priorityNodes))
+	copy(updatedPriorityNodes, priorityNodes)
```
Comment (Eslam-Nawara) [inline farmerbot/internal/farmerbot.go]: I think the copy and delete part should be replaced with empty slices and append nodes if exist.

File: farmerbot/internal/find_node_test.go [hunk 5]
```diff
@@ -67,65 +67,71 @@ func TestFindNode(t *testing.T) {
...
-		assert.Contains(t, farmerbot.nodes, nodeID)
 
-		farmerbot.addNode(oldNode1)
-		farmerbot.addNode(oldNode2)
+		_, node, err := farmerbot.getNode(nodeID)
+		assert.NoError(t, err)
+		assert.Contains(t, farmerbot.nodes, node)
+
+		assert.NoError(t, farmerbot.updateNode(oldNode1))
+		assert.NoError(t, farmerbot.updateNode(oldNode2))
```
Comment (Eslam-Nawara) [inline farmerbot/internal/find_node_test.go]: Why are we updating the `oldNodes`

Comment (rawdaGastan) [reply farmerbot/internal/find_node_test.go]: To reset changes made by `findNode` function

**Review by Eslam-Nawara — APPROVED** [review]

File: farmerbot/internal/farmerbot_test.go [hunk 6]
```diff
@@ -32,6 +32,7 @@ func TestFarmerbot(t *testing.T) {
 	inputs := Config{
 		FarmID:        1,
 		IncludedNodes: []uint32{1, 2},
+		PriorityNodes: []uint32{2},
```
Comment (xmonader) [inline farmerbot/internal/farmerbot_test.go:35]: I think we need multiple validations for priority nodes field e.g duplicated record, invalid node id, missing node id, node id that doesn't belong to a farm, .. etc

Comment (rawdaGastan) [reply farmerbot/internal/farmerbot_test.go:35]: To be added: 
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/1017

**Review by xmonader — APPROVED** [review]

**Pull request merged**